File: examples/models/donut.py
from __future__ import print_function

import base64
from math import pi, sin, cos

from bokeh.util.browser import view
from bokeh.colors import skyblue, seagreen, tomato, orchid, firebrick, lightgray
from bokeh.document import Document
from bokeh.embed import file_html
from bokeh.models.glyphs import Wedge, AnnularWedge, ImageURL, Text
from bokeh.models import ColumnDataSource, Plot, Range1d
from bokeh.resources import INLINE
from bokeh.sampledata.browsers import browsers_nov_2013, icons

df = browsers_nov_2013

xdr = Range1d(start=-2, end=2)
ydr = Range1d(start=-2, end=2)

plot = Plot(x_range=xdr, y_range=ydr, plot_width=800, plot_height=800)
plot.title.text = "Web browser market share (November 2013)"
plot.toolbar_location = None

colors = {"Chrome": seagreen, "Firefox": tomato, "Safari": orchid, "Opera": firebrick, "IE": skyblue, "Other": lightgray}

aggregated = df.groupby("Browser").agg(sum)
selected = aggregated[aggregated.Share >= 1].copy()
selected.loc["Other"] = aggregated[aggregated.Share < 1].sum()
browsers = selected.index.tolist()

radians = lambda x: 2*pi*(x/100)
angles = selected.Share.map(radians).cumsum()

end_angles = angles.tolist()
start_angles = [0] + end_angles[:-1]

browsers_source = ColumnDataSource(dict(
    start  = start_angles,
    end    = end_angles,
    colors = [colors[browser] for browser in browsers ],
))

glyph = Wedge(x=0, y=0, radius=1, line_color="white",
    line_width=2, start_angle="start", end_angle="end", fill_color="colors")
plot.add_glyph(browsers_source, glyph)

def polar_to_cartesian(r, start_angles, end_angles):
    cartesian = lambda r, alpha: (r*cos(alpha), r*sin(alpha))
    points = []

    for start, end in zip(start_angles, end_angles):
        points.append(cartesian(r, (end + start)/2))

    return zip(*points)

first = True

for browser, start_angle, end_angle in zip(browsers, start_angles, end_angles):
    versions = df[(df.Browser == browser) & (df.Share >= 0.5)]
    angles = versions.Share.map(radians).cumsum() + start_angle
    end = angles.tolist() + [end_angle]
    start = [start_angle] + end[:-1]
    base_color = colors[browser]
    fill = [ base_color.lighten(i*0.05) for i in range(len(versions) + 1) ]
    text = [ number if share >= 1 else "" for number, share in zip(versions.VersionNumber, versions.Share) ]
    x, y = polar_to_cartesian(1.25, start, end)

    source = ColumnDataSource(dict(start=start, end=end, fill=fill))
    glyph = AnnularWedge(x=0, y=0,
        inner_radius=1, outer_radius=1.5, start_angle="start", end_angle="end",
        line_color="white", line_width=2, fill_color="fill")
    plot.add_glyph(source, glyph)


    text_angle = [(start[i]+end[i])/2 for i in range(len(start))]
    text_angle = [angle + pi if pi/2 < angle < 3*pi/2 else angle for angle in text_angle]

    if first and text:
        text.insert(0, '(version)')
        offset = pi / 48
        text_angle.insert(0, text_angle[0] - offset)
        start.insert(0, start[0] - offset)
        end.insert(0, end[0] - offset)
        x, y = polar_to_cartesian(1.25, start, end)
        first = False

    text_source = ColumnDataSource(dict(text=text, x=x, y=y, angle=text_angle))
    glyph = Text(x="x", y="y", text="text", angle="angle",
        text_align="center", text_baseline="middle")
    plot.add_glyph(text_source, glyph)


def to_base64(png):
    return "data:image/png;base64," + base64.b64encode(png).decode("utf-8")

urls = [ to_base64(icons.get(browser, b"")) for browser in browsers ]
x, y = polar_to_cartesian(1.7, start_angles, end_angles)

icons_source = ColumnDataSource(dict(urls=urls, x=x, y=y))
glyph = ImageURL(url="urls", x="x", y="y", anchor="center")
plot.add_glyph(icons_source, glyph)

text = [ "%.02f%%" % value for value in selected.Share ]
x, y = polar_to_cartesian(0.7, start_angles, end_angles)

text_source = ColumnDataSource(dict(text=text, x=x, y=y))
glyph = Text(x="x", y="y", text="text", text_align="center", text_baseline="middle")
plot.add_glyph(text_source, glyph)

doc = Document()
doc.add_root(plot)

if __name__ == "__main__":
    doc.validate()
    filename = "donut.html"
    with open(filename, "w") as f:
        f.write(file_html(doc, INLINE, "Donut Chart"))
    print("Wrote %s" % filename)
    view(filename)

File: bokehjs/examples/donut/donut.ts
namespace WebBrowserMarketShare {
  import plt = Bokeh.Plotting;
  const {zip, unzip, sum, cumsum} = Bokeh.LinAlg;

  Bokeh.set_log_level("info");
  Bokeh.logger.info(`Bokeh ${Bokeh.version}`);

  function to_cartesian(r: number, alpha: number): [number, number] {
    return [r*Math.cos(alpha), r*Math.sin(alpha)]
  }

  function to_radians(x: number) {
    return 2*Math.PI*(x/100)
  }

  function read_csv_from(id: string) {
    const text = document.getElementById(id).innerHTML
    return text.split("\n")
         .map((line) => line.trim())
         .filter((line) => line.length > 0)
         .map((line) => line.split(/, /).map((val) => val.trim()))
  }

  interface MonthlyShares {
    year: number
    month: string
    browsers: string[]
    shares: number[]
  }

  const data: MonthlyShares[] = []

  let _browsers: string[] = null
  let year: number = null

  for (const [head, ...tail] of read_csv_from("data")) {
    const _year = parseInt(head)
    if (!isNaN(_year)) {
      [year, _browsers] = [_year, tail]
    } else {
      const month = head

      const shares = tail.map((val) => parseFloat(val))
      const browsers = _browsers.slice()

      for (let i = 0; i < shares.length;) {
        if (isNaN(shares[i])) {
          shares.splice(i, 1)
          browsers.splice(i, 1)
        } else
          i++
      }

      const other = 100 - sum(shares)
      if (other > 0) {
        browsers.push("Other")
        shares.push(other)
      }

      data.push({year: year, month: month, browsers: browsers, shares: shares})
    }
  }

  interface BrowserInfo {
    description?: string
    color?: Bokeh.Color
    icon?: string
  }

  const info: Bokeh.Map<BrowserInfo> = {
    Other: {color: "gray"},
  }

  for (const row of read_csv_from("info")) {
    const [browser, description, color, icon] = row
    info[browser] = {
      description: description,
      color: color,
      icon: icon,
    }
  }

  const fig = plt.figure({
    x_range: [-2, 2], y_range: [-2, 2],
    width: 600, height: 600,
    x_axis_type: null, y_axis_type: null,
    tools: [], toolbar_location: null,
  })

  function render(item: MonthlyShares) {
    fig.title = `${item.month} ${item.year}`

    const source = new Bokeh.ColumnDataSource({
      data: {
        names: item.browsers,
        shares: item.shares,
      },
    })

    const end_angles = cumsum(item.shares.map(to_radians))
    const start_angles = [0].concat(end_angles.slice(0, -1))
    const half_angles = zip(start_angles, end_angles).map(([start, end]) => (start + end)/2)
    const colors = item.browsers.map((name) => info[name].color)

    fig.wedge({x: 0, y: 0, radius: 1.5, source: source,
         start_angle: start_angles, end_angle: end_angles,
         line_color: "white", line_width: 1, fill_color: colors})

    const icons = item.browsers.map((name) => info[name].icon)
    const [x0, y0] = unzip(half_angles.map((angle) => to_cartesian(1.7, angle)))
    fig.image_url(icons, x0, y0, null, null, {source: source, anchor: "center"})

    const texts = item.shares.map((share) => {
      if (share <= 2.0)
        return null
      else
        return Bokeh.sprintf("%.02f%%", share)
    })
    const text_angles = item.shares.map((share, i) => {
      if (share <= 5.0)
        return half_angles[i]
      else
        return 0
    })
    const [x1, y1] = unzip(half_angles.map((angle) => to_cartesian(1.0, angle)))
    fig.text(x1, y1, texts, {source: source, angle: text_angles, text_align: "center", text_baseline: "middle"})
  }

  const tap = new Bokeh.TapTool({
    behavior: "inspect",
    callback: (ds) => {
      if (!paused) {
        const cds = ds as Bokeh.ColumnDataSource
        const i = cds.inspected['1d'].indices[0]
        const name = cds.data["names"][i]
        const share = Bokeh.sprintf("%.02f%%", cds.data["shares"][i])
        fig.title = `${name}: ${share}`
      }
      paused = !paused
    },
  })
  fig.add_tools(tap)

  let index = data.length-1
  let paused = false

  setInterval(() => {
    if (!paused) {
      fig.renderers = fig.renderers.filter((r) => !(r instanceof Bokeh.GlyphRenderer))
      render(data[index])
      if (--index < 0)
        index = data.length-1
    }
  }, 1000)

  plt.show(fig)
}
